examples/bsearch.masm

```diff
@@ -0,0 +1,208 @@
+#!            `val` is value to search.
+#! Output: [found, index] where `found` is searching result (`true`/`false`),
+#!         index `[0, count)` of `val`, if `val` was found; otherwise index in array, where `val` should be inserted.
+proc.binary_search_stack
+    push.STARTING_MEMORY_ADDRESS            # [STARTING_MEMORY_ADDRESS, count, arr[0], arr[1], ..., arr[count - 1], val, ...]
+    exec.read_stack_to_memory               # [end, val, ...]
+
+    push.STARTING_MEMORY_ADDRESS            # [start, end, val, ...], start = STARTING_MEMORY_ADDRESS
+    exec.binary_search                      # [found, addr, ...]
+
+    swap                                    # [addr, found, ...]
+    push.STARTING_MEMORY_ADDRESS            # [STARTING_MEMORY_ADDRESS, addr, found, ...]
+    u32wrapping_sub                         # [index, found, ...], index = addr - STARTING_MEMORY_ADDRESS
+    swap                                    # [found, index, ...]
+end
+
+proc.test_empty
+    push.5.0
+    exec.binary_search_stack
+    assertz     # Expect "not found"
+    assertz     # Expect insertion index is `0`
+end
+
+proc.test_solid
+    push.4.5.4.3.2.1.5
+    exec.binary_search_stack
+    assert      # Expect "found"
+    push.3
+    assert_eq   # Expect index is `3`
+end
+
+proc.test_first
+    push.12.50.44.36.23.12.5
+    exec.binary_search_stack
+    assert      # Expect "found"
+    push.0
+    assert_eq   # Expect index is `0`
+end
+
+proc.test_last
+    push.50.50.44.36.23.12.5
+    exec.binary_search_stack
+    assert      # Expect "found"
+    push.4
+    assert_eq   # Expect index is `4`
+end
+
+proc.test_gaps_right_found
+    push.44.50.44.36.23.12.5
+    exec.binary_search_stack
+    assert      # Expect "found"
+    push.3
+    assert_eq   # Expect index is `3`
+end
+
+proc.test_gaps_left_found
+    push.23.50.44.36.23.12.5
+    exec.binary_search_stack
+    assert      # Expect "found"
+    push.1
+    assert_eq   # Expect index is `1`
+end
+
+proc.test_gaps_right_not_found
+    push.40.50.44.36.23.12.5
+    exec.binary_search_stack
+    assertz     # Expect "not found"
+    push.3
+    assert_eq   # Expect insertion index is `3`
+end
+
+proc.test_gaps_left_not_found
+    push.30.50.44.36.23.12.5
+    exec.binary_search_stack
+    assertz     # Expect "not found"
+    push.2
+    assert_eq   # Expect insertion index is `2`
+end
+
+#! Runs tests.
+proc.run_tests
+    call.test_empty
+    call.test_solid
+    call.test_first
+    call.test_last
+    call.test_gaps_right_found
+    call.test_gaps_left_found
+    call.test_gaps_right_not_found
+    call.test_gaps_left_not_found
+end
+
+begin
+    dup                                     # [count, count, ...]
+    neq.0                                   # [count != 0, count, ...]
+    if.true                                 # [count, arr[0], arr[1], ..., arr[count - 1], val, ...]
+        exec.binary_search_stack            # [found, index]
```

Comment:
can you maybe get rid of the `found` in the final output?

You could return all `0` if you didn't find the value and the pos of the stack if you found it

Reply:
I think, it wouldn't be a good idea to return indexes of the stack. We work with index in array, which is passed through the stack. If value is not found, `index == 0` would conflict with the index of 0th element. I tried to use special index like `0xFFFFFFFF` for such cases, but it made the code worse.

Reply:
agreed, thanks for the explanation